write-fonts/src/tables/gvar.rs

```diff
@@ -325,22 +325,41 @@ impl GlyphDelta {
     }
 }
 
+pub struct AxisCoordinates {
+    peak: F2Dot14,
+    intermediates: Option<(F2Dot14, F2Dot14)>,
+}
+
+impl AxisCoordinates {
+    /// Get the minimum and maximum only if they differ from their defaults as
+    /// calculated from the peak.
+    fn required_intermediates(&self) -> Option<(F2Dot14, F2Dot14)> {
+        self.intermediates.filter(|(min, max)| {
+            (min, max) != (&self.peak.min(F2Dot14::ZERO), &self.peak.max(F2Dot14::ZERO))
+        })
+    }
+}
+
 impl GlyphDeltas {
     /// Create a new set of deltas.
     ///
     /// A None delta means do not explicitly encode, typically because IUP suggests
     /// it isn't required.
-    pub fn new(
-        peak_tuple: Tuple,
-        deltas: Vec<GlyphDelta>,
-        intermediate_region: Option<(Tuple, Tuple)>,
-    ) -> Self {
-        if let Some((start, end)) = intermediate_region.as_ref() {
-            assert!(
-                start.len() == end.len() && start.len() == peak_tuple.len(),
-                "all tuples must have equal length"
-            );
-        }
+    pub fn new(axis_coordinates: Vec<AxisCoordinates>, deltas: Vec<GlyphDelta>) -> Self {
+        let peak_tuple = Tuple::new(axis_coordinates.iter().map(|coords| coords.peak).collect());
+
+        // File size optimisation: only write explicit minimums and maximums if
+        // at least one of the tents requires them.
+        // https://github.com/fonttools/fonttools/blob/b467579c/Lib/fontTools/ttLib/tables/TupleVariation.py#L184-L193
+        let required_intermediates = axis_coordinates
+            .iter()
+            .map(|coords| coords.required_intermediates())
+            .collect::<Option<Vec<_>>>();
```

Comment:
we should be able to get rid of this allocation and do this all in one iter chain? I think? 🤔

Reply:
This did indeed turn out to be elegant and possible, with this neat new addition: https://users.rust-lang.org/t/unzip-with-error-handling/110250/2

...but I then realised that when reimplementing in fontations I had accidentally switched the semantics from `any(...)` to `all(...)`, and so unfortunately it all had to be scrapped again ☔

With the intention corrected, there doesn't seem to be a better semantic match than a mutable fold, which @RickyDaMa and I have now implemented, although it does at least stay readable and keep allocations to a minimum.

In general, representing a combination of `collect(...)` and `any(...)` in an iter chain has turned out to be a bit of an abstraction tar-pit; each time you follow an improvement, you get bumped back into the same narrow path :(